RFC-0037-Interoperability-Standard-of-3rd-Backend-Integration-Mechanism.md

```diff
@@ -0,0 +1,91 @@
+| Refactoring | ​​Use DeviceIndex instead of int in CUDA wrappers, derived class’ override function is not modified. | 1 | [#119142](https://github.com/pytorch/pytorch/pull/119142) |
+| Refactoring | Move new trace utils from source to header, which leads to some symbols can’t be found. | 1 | [#114367](https://github.com/pytorch/pytorch/pull/114367/files) |
+| Refactoring | Migrate to getCvar* functions for env variable checking, which leads to function name can’t be found. | 1 | [#113797](https://github.com/pytorch/pytorch/pull/113797) |
+| New Features | Add support for new data types, data type assert fails. | 2 | [#107586](https://github.com/pytorch/pytorch/pull/107586), [#116594](https://github.com/pytorch/pytorch/pull/116594) |
+| New Features | Add function to materialize COW storages, which add a pure virtual function Allocator::copy_data, derived class didn’t implement this pure virtual function. | 2 | [#117053](https://github.com/pytorch/pytorch/pull/117053), [#113396](https://github.com/pytorch/pytorch/pull/113396) |
+
+Based on the above table, it is evident that upstream modifications do have a negative impact on out-of-tree third-party devices. These modifications directly affect the functionality of third-party devices and greatly increase the cost of adaptation and maintenance.
+
+
+### **Standardized Integration of Third-party Device**
+PyTorch is a complex framework with numerous features, making it a significant challenge for third-party devices to adapt to. The quality of integration methods directly impacts the stability and sustainability of the subsequent functionality of third-party backends. Therefore, standardizing the integration for third-party devices not only reduces integration barriers and time costs but also enhances compatibility and stability. This, in turn, solidifies PyTorch's leading position in the field of artificial intelligence frameworks.
+
+
+## **Proposed Implementation**
+![Architecture Outline](./RFC-0030-assets/3rd_backend_architecture.png)
+
+As shown in the above diagram, based on the third-party device integration mechanism (PrivateUse1) and a real device, a in-tree and lightweight backend will be added. This backend synchronously adds minimal test suites and end-to-end integration documentation. It serves as both a quality assurance for the third-party device integration mechanism and a standard official reference implementation for third-party devices. This backend has the following characteristics:
+
+* **Functionality**: Implements only the most essential and minimal functionality while covering the entire feature set.
+* **Tests**: Minimal test suite
+* **Usage**: Reserved for CI/CD and official standard reference implementation, not intended for user use.
+* **Compilation**: Separate compilation entry point, separate from other backends.
+* **Independence**: The related code is stored separately, without coupling with other functional modules of PyTorch.
```

Comment:
I feel like a lot of this can be achieved by making this demo backend another repo on the pytorch/* org that we take as a submodule for testing only and we build. This way, it is:
- Fully independent codebase from core, just like the real out-of-tree backends
- A real end-to-end example of how to make a new backend
- Can be fully tested in our CI and be pinned as needed

Reply:
Good idea.
Compared to in-tree, this is indeed a better way, thank you.

Reply:
In addition, compared with in-tree, out-of-tree will be more troublesome when merging some PRs that cause the 3rd device to be unavailable, because developers need to fix the version between the two repos or other methods. coordination.

Reply:
Yes there are pros and cons both ways. Looking into what would be the easiest to setup. Will get back to you shortly.